datafusion/expr/src/expr_schema.rs

```diff
@@ -173,8 +173,30 @@ impl ExprSchemable for Expr {
             Expr::Alias(expr, _)
             | Expr::Not(expr)
             | Expr::Negative(expr)
-            | Expr::Sort(Sort { expr, .. })
-            | Expr::InList(InList { expr, .. }) => expr.nullable(input_schema),
+            | Expr::Sort(Sort { expr, .. }) => expr.nullable(input_schema),
+
+            Expr::InList(InList { expr, list, .. }) => {
+                // Avoid inspecting too many expressions.
+                const MAX_INSPECT_LIMIT: usize = 6;
```

Comment:
I want to know why use this number? Is it the practice from other systems?

Reply:
spark handle it simplified.

```scala
override def nullable: Boolean = children. Exists(_.nullable)
```

Reply:
> I want to know why use this number? Is it the practice from other systems?

@jackwener No. The `nullable` function may be called multiple times during the optimization phase, 
So I think adding a limitation would be preferable in order to prevent it from being excessively slow.
But I'm not quite sure what would be an appropriate number.

   

> spark handle it simplified.
> 
> ```scala
> override def nullable: Boolean = children. Exists(_.nullable)
> ```
This seems to be a cache style.
We can implement this by precomputing nullable in the `InList::new()` function.
But the disadvantage are：
- We need a new field for the `InList` struct
- The precomputed nullable may not be used.

@jackwener  Which solution do you prefer?

Reply:
As @jonahgao  hints at , trying to add some sort of cache / memoization is going to be challenging given Rust and how DataFusion is structured

I think the current PR solution is good:
1. It is well tested 
2. It is conservative (it might say an InList is nullable that isn't, but I think that will not generate wrong results, only potentially less optimal plans)

Thus I think we should merge this PR as is and then as a follow on PR we can remove the limit or increase it, etc.

Reply:
thanks @jonahgao @alamb